fastcrypto-zkp/src/api.rs

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use std::io::Cursor;
+
+use ark_bls12_381::Bls12_381;
+use ark_groth16::{Proof, VerifyingKey};
+use ark_relations::r1cs::SynthesisError;
+use ark_serialize::CanonicalDeserialize;
+
+use crate::conversions::BlsFr;
+use crate::verifier::{process_vk_special, verify_with_processed_vk};
+
+/// Verify Groth16 proof using the serialized representation of verifying key, input and proof.
+/// This assumes canoncical ser/de defined in Arkworks, instead of the Zcash bytes representation.
+pub fn verify_groth16_in_bytes(
```

Comment:
I don't quite know how you intend to use this, but just to make clarify: the verification of a Groth16 proof is inherently broken up in two phases:
1. setting up the assets of the verifying key: this involves computation (one pairing e(P, Q)), and "binds" the verification to one particular proof statement (e.g. "I know a sha-256 pre-image that ends in `0xdeadbeef`"),
2. verifying one proof instance: which also involves computation, obviously, e.g. "those points prove I have a valid pre-image, which first 3 bits are public and passed as input".

The expectation is that a smart contract setting up the proof would call 1. once, and then call 2. repeatedly for many proof instances.

Here, the risk with presenting a single API is that the pre-processing of the verifying key contains computation that should be mutualized across the proof instances. If it is not, the price of verification will be one more pairing than necessary per instance.